src/cmake/CMakeBasics.cmake

```diff
@@ -19,20 +19,35 @@ include(cmake/AxomMacros.cmake)
 #------------------------------------------------------------------------------
 include(cmake/thirdparty/SetupAxomThirdParty.cmake)
 
-if(NOT CMAKE_CONFIGURATION_TYPES)
-    #--------------------------------------------------------------------------
-    # Add define we can use when debug builds are enabled
-    #--------------------------------------------------------------------------
-    if( CMAKE_BUILD_TYPE MATCHES "(Debug|RelWithDebInfo)" )
-        add_definitions(-DAXOM_DEBUG)
-    endif()
-else ()
-    set_property(DIRECTORY APPEND PROPERTY COMPILE_DEFINITIONS
-      $<$<CONFIG:Debug>:AXOM_DEBUG>
-      $<$<CONFIG:RelWithDebInfo>:AXOM_DEBUG>
-    )
+#------------------------------------------------------------------------------
+# Set up AXOM_DEBUG compiler define string, as appropriate, based on config type.
+# Result is stored in AXOM_DEBUG_DEFINE_STRING cache variable
+#------------------------------------------------------------------------------
+set(AXOM_DEBUG_DEFINE_STRING "")
+
+# Handle the three valid values for AXOM_DEBUG_DEFINE: {on, off, default}
+string(TOUPPER "${AXOM_DEBUG_DEFINE}" _axom_debug_define_upper)
+if("${_axom_debug_define_upper}" MATCHES "ON|TRUE")
+  set(AXOM_DEBUG_DEFINE_STRING "AXOM_DEBUG")
+elseif("${_axom_debug_define_upper}" MATCHES "OFF|FALSE")
+  # no-op
+elseif("${_axom_debug_define_upper}" MATCHES "DEFAULT")
+  # Default behavior is to be on for Debug and RelWithDebInfo configurations and off otherwise
+  if(NOT CMAKE_CONFIGURATION_TYPES)
+      if( CMAKE_BUILD_TYPE MATCHES "(Debug|RelWithDebInfo)" )
+        set(AXOM_DEBUG_DEFINE_STRING "AXOM_DEBUG")
+      endif()
+  else ()
+      set(AXOM_DEBUG_DEFINE_STRING "$<$<CONFIG:Debug,RelWithDebInfo>:AXOM_DEBUG>")
+  endif()
+else()  # Handle bad value for AXOM_DEBUG_DEFINE config variable
+  message(FATAL_ERROR 
+    "Invalid value for AXOM_DEBUG_DEFINE. Must be 'DEFAULT', 'ON' or 'OFF'; was '${AXOM_DEBUG_DEFINE}'")
 endif()
 
+set(AXOM_DEBUG_DEFINE_STRING "${AXOM_DEBUG_DEFINE_STRING}" CACHE STRING "" FORCE)
+mark_as_advanced(AXOM_DEBUG_DEFINE_STRING)
```

Comment:
This chunk handles if we define the `AXOM_DEBUG` define for single and multi-config generators